### PR TITLE
Implement SpotlightSelector component for RankingsOverlay

### PR DESCRIPTION
**Pull request by EVAST9919**

### CODE REVIEW

File: osu.Game/Online/API/Requests/Responses/APISpotlight.cs [hunk 1]
```diff
@@ -0,0 +1,34 @@
...
+        public bool ModeSpecific;
+
+        [JsonProperty(@"start_date")]
+        public DateTimeOffset StartDate;
+
+        [JsonProperty(@"end_date")]
+        public DateTimeOffset EndDate;
+
+        [JsonProperty(@"participant_count")]
+        public int? ParticipiantCount;
```
Comment (bdach) [inline osu.Game/Online/API/Requests/Responses/APISpotlight.cs]:
```suggestion
        public int? ParticipantCount;
```

File: osu.Game/Online/API/Requests/GetSpotlightsRequest.cs [hunk 2]
```diff
@@ -0,0 +1,20 @@
...
+using osu.Game.Online.API.Requests.Responses;
+
+namespace osu.Game.Online.API.Requests
+{
+    public class GetSpotlightsRequest : APIRequest<SpotlightsArray>
+    {
+        protected override string Target => "spotlights";
+    }
+
+    public class SpotlightsArray
```
Comment (bdach) [inline osu.Game/Online/API/Requests/GetSpotlightsRequest.cs]: Sort of misleading name, I'd go for
```suggestion
    public class SpotlightsCollection
```

File: osu.Game/Overlays/Rankings/SpotlightSelector.cs [hunk 3]
```diff
@@ -0,0 +1,165 @@
...
+
+        private void onSelectionChanged(ValueChangedEvent<APISpotlight> spotlight)
+        {
+            startDateColumn.Value = dateToString(spotlight.NewValue.StartDate);
+            endDateColumn.Value = dateToString(spotlight.NewValue.EndDate);
+        }
+
+        private string dateToString(DateTimeOffset date) => $"{date.Year}-{date.Month:D2}-{date.Day:D2}";
+
+        private class InfoCoulmn : FillFlowContainer
```
Comment (bdach) [inline osu.Game/Overlays/Rankings/SpotlightSelector.cs]:
```suggestion
        private class InfoColumn : FillFlowContainer
```

File: osu.Game/Overlays/Rankings/SpotlightSelector.cs [hunk 4]
```diff
@@ -0,0 +1,165 @@
...
+            api.Queue(request);
+        }
+
+        private void onSelectionChanged(ValueChangedEvent<APISpotlight> spotlight)
+        {
+            startDateColumn.Value = dateToString(spotlight.NewValue.StartDate);
+            endDateColumn.Value = dateToString(spotlight.NewValue.EndDate);
+        }
+
+        private string dateToString(DateTimeOffset date) => $"{date.Year}-{date.Month:D2}-{date.Day:D2}";
```
Comment (bdach) [inline osu.Game/Overlays/Rankings/SpotlightSelector.cs]: That's a very roundabout way of writing
```suggestion
        private string dateToString(DateTimeOffset date) => date.ToString("yyyy-MM-dd");
```

**Review by bdach — CHANGES_REQUESTED** [review]
Visually the dropdown header text could use more padding from the sides, although I do know that it's not something immediately accessible or modifiable.

**Review by bdach — DISMISSED** [review]
I don't think I have any issues with this at its current state anymore.

File: osu.Game/Overlays/Rankings/SpotlightSelector.cs [hunk 5]
```diff
@@ -0,0 +1,165 @@
...
+            background.Colour = colours.GreySeafoam;
+        }
+
+        protected override void LoadComplete()
+        {
+            base.LoadComplete();
+            SelectedSpotlight.BindValueChanged(onSelectionChanged);
+        }
+
+        public void FetchSpotlights()
```
Comment (smoogipoo) [inline osu.Game/Overlays/Rankings/SpotlightSelector.cs]: The selector should not be doing the web request. It should:
1. Take in a set of spotlights.
2. Handle selecting a spotlight.
3. Return a spotlight.

The web request would be part of the overlay, and in this case the test scene providing the spotlights (the test scene should provide a custom set of spotlights).

**Review by smoogipoo — CHANGES_REQUESTED** [review]

File: osu.Game/Overlays/Rankings/SpotlightSelector.cs (hunk 5 above)
Comment (EVAST9919) [reply osu.Game/Overlays/Rankings/SpotlightSelector.cs]: Got it

**Comment by smoogipoo** [comment]
Not sure where you got `participant_count` from but it's not returned by the API. It, along with the maps, will be returned in a separate request.

I've updated code to reflect these changes (remove said columns), and made a few refactorings.

@peppy quick check plz

**Comment by EVAST9919** [comment]
@smoogipoo it's true that this will be returned with a separate request, but it's still a part of the spotlight
Screenshot_1

**Review by bdach — DISMISSED** [review]
Looks better to me for sure. I kind of got used to some test scenes being tightly coupled to API at this point, but I shouldn't have done so as this is the more correct approach.

**Review by bdach — APPROVED** [review]

**Review by smoogipoo — APPROVED** [review]
Seems good

**Pull request merged**